packages/expo-updates/ios/EXUpdates/EXUpdatesErrorRecovery.m

```diff
@@ -0,0 +1,220 @@
+      @(EXUpdatesErrorRecoveryTaskLaunchCached),
+      @(EXUpdatesErrorRecoveryTaskLaunchEmbedded),
+      @(EXUpdatesErrorRecoveryTaskCrash)
+    ].mutableCopy;
+    _isRunning = NO;
+    _isWaitingForRemoteUpdate = NO;
+    _errorRecoveryQueue = errorRecoveryQueue;
+    _remoteLoadTimeout = remoteLoadTimeout;
+    _encounteredErrors = [NSMutableArray new];
+  }
+  return self;
+}
+
+- (void)handleError:(NSError *)error fromLaunchedUpdate:(nullable EXUpdatesUpdate *)launchedUpdate
+{
+  if (launchedUpdate) {
+    [self _markUpdateFailed:launchedUpdate];
+  }
+  [self _startPipelineWithEncounteredError:error];
+}
+
+- (void)handleException:(NSException *)exception fromLaunchedUpdate:(nullable EXUpdatesUpdate *)launchedUpdate
+{
+  if (launchedUpdate) {
+    [self _markUpdateFailed:launchedUpdate];
+  }
+  [self _startPipelineWithEncounteredError:exception];
+}
+
+- (void)notifyNewRemoteLoadStatus:(EXUpdatesRemoteLoadStatus)newStatus
+{
+  dispatch_async(_errorRecoveryQueue, ^{
+    if (!self->_isWaitingForRemoteUpdate) {
+      return;
+    }
+    self->_isWaitingForRemoteUpdate = NO;
+    if (newStatus == EXUpdatesRemoteLoadStatusNewUpdateLoaded) {
+      [self->_pipeline insertObject:@(EXUpdatesErrorRecoveryTaskLaunchNew) atIndex:0];
+    }
+    [self _runNextTask];
+  });
+}
+
+# pragma mark - pipeline tasks
+
+- (void)_startPipelineWithEncounteredError:(id)encounteredError
+{
+  dispatch_async(_errorRecoveryQueue, ^{
+    [self->_encounteredErrors addObject:encounteredError];
+    if (!self->_isRunning) {
+      self->_isRunning = YES;
+      [self _runNextTask];
+    }
+  });
+}
+
+- (void)_runNextTask
+{
+  dispatch_assert_queue(_errorRecoveryQueue);
+  NSNumber *nextTask = _pipeline[0];
+  [_pipeline removeObjectAtIndex:0];
+  switch ((EXUpdatesErrorRecoveryTask)nextTask.integerValue) {
+    case EXUpdatesErrorRecoveryTaskWaitForRemoteUpdate:
+      [self _waitForRemoteLoaderToFinish];
+      break;
+    case EXUpdatesErrorRecoveryTaskLaunchNew:
+    case EXUpdatesErrorRecoveryTaskLaunchCached:
+      [self _tryRelaunchFromCache];
+      break;
+    case EXUpdatesErrorRecoveryTaskLaunchEmbedded:
+      [self _tryLaunchEmbeddedUpdate];
+      break;
+    case EXUpdatesErrorRecoveryTaskCrash:
+      [self _crash];
+      break;
+    default:
+      @throw [NSException exceptionWithName:NSInternalInconsistencyException reason:@"EXUpdatesErrorRecovery cannot run the next task" userInfo:@{@"nextTaskValue": nextTask}];
+  }
+}
+
+- (void)_markUpdateFailed:(EXUpdatesUpdate *)update
+{
+  dispatch_async(_delegate.database.databaseQueue, ^{
+    NSError *error;
+    [self.delegate.database markUpdateFailed:update error:&error];
+    if (error) {
+      NSLog(@"Unable to mark update as failed in the local DB: %@", error.localizedDescription);
+      dispatch_async(self->_errorRecoveryQueue, ^{
+        [self->_pipeline removeObject:@(EXUpdatesErrorRecoveryTaskLaunchCached)];
+      });
+    }
+  });
+}
+
+- (void)_waitForRemoteLoaderToFinish
```

Comment:
It seems like this function is supposed to get called twice on a successful remote load (otherwise we would never add `EXUpdatesErrorRecoveryTaskLaunchNew` to the pipeline) but I can't seem to find the code where this happens. What am I missing?

Reply:
the way I have it set up is that if we call this function and a remote load is in progress, we then wait for `notifyNewRemoteLoadStatus` to be called when it's finished (which takes care of adding the task to the pipeline).